File: buzzvil-sdk-v6-ios/buzzvil-sdk-ios-objc/ViewController.m
#import "ViewController.h"
#import "UIButton+Custom.h"
#import "Native/NativeViewController.h"
#import "Native/Carousel/CarouselViewController.h"
#import "Interstitial/InterstitialViewController.h"
#import "BenefitHub/BenefitHubViewController.h"
#import "BenefitHub/BenefitHubContainerViewController.h"
#import "BenefitHub/BenefitHubNavigationCustomContainerViewController.h"
#import "Banner/BannerViewController.h"

@import BuzzvilSDK;

@interface ViewController ()

@property (nonatomic, strong, readonly) UIStackView *rootStackView;
@property (nonatomic, strong, readonly) UIButton *nativeButton;
@property (nonatomic, strong, readonly) UIButton *carouselButton;
@property (nonatomic, strong, readonly) UIButton *interstitialButton;
@property (nonatomic, strong, readonly) UIButton *benefitHubButton;
@property (nonatomic, strong, readonly) UIButton *benefitHubContainerButton;
@property (nonatomic, strong, readonly) UIButton *benefitHubNaviCustomContainerButton;
@property (nonatomic, strong, readonly) UIButton *bannerButton;
@property (nonatomic, strong, readonly) UIButton *inquiryButton;
@property (nonatomic, strong, readonly) UILabel *privacyConsentStatusLabel;
@property (nonatomic, strong, readonly) UIButton *loadPrivacyConsentStatusButton;
@property (nonatomic, strong, readonly) UIButton *grantPrivacyConsentButton;

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  [self setupView];
  [self setupLayout];
  [self setupEvent];
}

#pragma mark - UI setup
- (void)setupView {
  self.view.backgroundColor = [UIColor systemBackgroundColor];
  self.navigationItem.title = @"BuzzvilSDK-ObjectiveC";
  
  _rootStackView = [[UIStackView alloc] initWithFrame:CGRectZero];
  [self.view addSubview:_rootStackView];
  _rootStackView.axis = UILayoutConstraintAxisVertical;
  _rootStackView.spacing = 12;
  _rootStackView.distribution = UIStackViewDistributionFillEqually;
  
  _nativeButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_nativeButton setTitle:@"Native" forState:UIControlStateNormal];
  [_nativeButton applyCustomStyle];
  
  _interstitialButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_interstitialButton setTitle:@"Interstitial" forState:UIControlStateNormal];
  [_interstitialButton applyCustomStyle];
  
  _carouselButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_carouselButton setTitle:@"Carousel" forState:UIControlStateNormal];
  [_carouselButton applyCustomStyle];
  
  _benefitHubButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_benefitHubButton setTitle:@"BenefitHub" forState:UIControlStateNormal];
  [_benefitHubButton applyCustomStyle];
  
  _benefitHubContainerButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_benefitHubContainerButton setTitle:@"BenefitHub Container" forState:UIControlStateNormal];
  [_benefitHubContainerButton applyCustomStyle];
  
  _benefitHubNaviCustomContainerButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_benefitHubNaviCustomContainerButton setTitle:@"BenefitHub navi custom Container" forState:UIControlStateNormal];
  [_benefitHubNaviCustomContainerButton applyCustomStyle];

  _bannerButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_bannerButton setTitle:@"Banner" forState:UIControlStateNormal];
  [_bannerButton applyCustomStyle];
  
  _inquiryButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_inquiryButton setTitle:@"Inquiry" forState:UIControlStateNormal];
  [_inquiryButton applyCustomStyle];
  
  _privacyConsentStatusLabel = [[UILabel alloc] initWithFrame:CGRectZero];
  _privacyConsentStatusLabel.text = @"UNKNWON_STATUS";
  _privacyConsentStatusLabel.numberOfLines = 0;
  
  _loadPrivacyConsentStatusButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_loadPrivacyConsentStatusButton setTitle:@"Load PrivacyConsent Status" forState:UIControlStateNormal];
  [_loadPrivacyConsentStatusButton applyCustomStyle];
  
  _grantPrivacyConsentButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [_grantPrivacyConsentButton setTitle:@"Grant PrivacyConsent" forState:UIControlStateNormal];
  [_grantPrivacyConsentButton applyCustomStyle];

  [_rootStackView addArrangedSubview: _nativeButton];
  [_rootStackView addArrangedSubview: _interstitialButton];
  [_rootStackView addArrangedSubview: _carouselButton];
  [_rootStackView addArrangedSubview: _benefitHubButton];
  [_rootStackView addArrangedSubview: _benefitHubContainerButton];
  [_rootStackView addArrangedSubview: _benefitHubNaviCustomContainerButton];
  [_rootStackView addArrangedSubview: _bannerButton];
  [_rootStackView addArrangedSubview: _inquiryButton];
  [_rootStackView addArrangedSubview: _privacyConsentStatusLabel];
  [_rootStackView addArrangedSubview: _loadPrivacyConsentStatusButton];
  [_rootStackView addArrangedSubview: _grantPrivacyConsentButton];
}

- (void) setupLayout {
  _rootStackView.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [_rootStackView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
    [_rootStackView.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor],
    [_rootStackView.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor],
    [_rootStackView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor],
  ]];
}


- (void)setupEvent {
  [_benefitHubButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushBenefitHubViewController:)]];
  [_nativeButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushNativeViewController:)]];
  [_carouselButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushCarouselViewController:)]];
  [_interstitialButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushInterstitialViewController:)]];
  [_benefitHubContainerButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushBenefitHubContainerViewController:)]];
  [_benefitHubNaviCustomContainerButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushBenefitHubNaviCustomContainerViewController:)]];
  [_bannerButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushBannerViewController:)]];
  [_inquiryButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(showInquiryPage:)]];
  [_loadPrivacyConsentStatusButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(loadPrivacyConsentStatus:)]];
  [_grantPrivacyConsentButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(grantPribacyConsent:)]];
}

- (void)pushBenefitHubViewController:(id)sender {
  BenefitHubViewController *benefitHubViewController = [[BenefitHubViewController alloc] init];
  [self.navigationController pushViewController:benefitHubViewController animated:YES];
}

- (void)pushNativeViewController:(id)sender {
  NativeViewController *nativeViewController = [[NativeViewController alloc] init];
  [self.navigationController pushViewController:nativeViewController animated:YES];
}

- (void)pushCarouselViewController:(id)sender {
  CarouselViewController *carouselViewController = [[CarouselViewController alloc] init];
  [self.navigationController pushViewController:carouselViewController animated:YES];
}

- (void)pushInterstitialViewController:(id)sender {
  InterstitialViewController *interstitialViewController = [[InterstitialViewController alloc] init];
  [self.navigationController pushViewController:interstitialViewController animated:YES];
}

- (void)pushBenefitHubContainerViewController:(id)sender {
  BenefitHubContainerViewController *benefitHubContainerViewController = [[BenefitHubContainerViewController alloc] init];
  [self.navigationController pushViewController:benefitHubContainerViewController animated:YES];
}

- (void)pushBenefitHubNaviCustomContainerViewController:(id)sender {
  BenefitHubNavigationCustomContainerViewController *benefitHubNaviCustomConatiner = [[BenefitHubNavigationCustomContainerViewController alloc] init];
  [self.navigationController pushViewController:benefitHubNaviCustomConatiner animated:YES];
}

- (void)pushBannerViewController:(id)sender {
  BannerViewController *bannerViewController = [[BannerViewController alloc] init];
  [self.navigationController pushViewController:bannerViewController animated:YES];
}

- (void)showInquiryPage:(id)sender {
  [[BuzzAdBenefit sharedInstance] openInquiryPageWithUnitId:@"YOUR_UNIT_ID"];
}

- (void)loadPrivacyConsentStatus:(id)sender {
  __weak typeof(self) weakSelf = self;
  
  [BuzzAdBenefit.sharedInstance loadPrivacyConsentStatusOnSuccess:^(enum BuzzPrivacyConsentStatus status) {
    __strong typeof(self) strongSelf = weakSelf;
    switch (status) {
      case BuzzPrivacyConsentStatusGranted:
        strongSelf.privacyConsentStatusLabel.text = @"Granted";
        break;
      case BuzzPrivacyConsentStatusRevoked:
        strongSelf.privacyConsentStatusLabel.text = @"Revoked";
        break;
    }
  } onFailure:^(NSError * _Nonnull error) {
    __strong typeof(self) strongSelf = weakSelf;
    
    strongSelf.privacyConsentStatusLabel.text = [NSString stringWithFormat:@"LoadPrivacyConsentStatus is failed Error: %@", error.localizedDescription];
  }];
}

- (void)grantPribacyConsent:(id)sender {
  __weak typeof(self) weakSelf = self;
  
  [BuzzAdBenefit.sharedInstance grantPrivacyConsentOnSuccess:^{
    __strong typeof(self) strongSelf = weakSelf;
    strongSelf.privacyConsentStatusLabel.text = @"Granted";
  } onFailure:^(NSError * _Nonnull error) {
    __strong typeof(self) strongSelf = weakSelf;
    strongSelf.privacyConsentStatusLabel.text = [NSString stringWithFormat:@"GrantPrivacyPolicy is failed Error: %@", error.localizedDescription];
  }];
}

@end

File: buzzvil-sdk-v6-ios/buzzvil-sdk-ios-swift/ViewController.swift
import UIKit
import BuzzvilSDK

class ViewController: UIViewController {
  private lazy var rootStackView: UIStackView = {
    let stackView = UIStackView(frame: .zero)
    stackView.axis = .vertical
    stackView.spacing = 12
    stackView.distribution = .fillEqually
    return stackView
  }()
  
  private lazy var nativeButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Native", for: .normal)
    button.addTarget(self, action: #selector(pushNativeAdViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var interstitialButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Interstitial", for: .normal)
    button.addTarget(self, action: #selector(pushInterstitialViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var carouselButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Carousel", for: .normal)
    button.addTarget(self, action: #selector(pushCarouselViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var benefitHubButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("BenefitHub", for: .normal)
    button.addTarget(self, action: #selector(pushBenefitHubViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var benefitHubContainerButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("BenefitHub Container", for: .normal)
    button.addTarget(self, action: #selector(pushBenefitHubContainerViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var benefitHubCustomNaviContainerButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("BenefitHub CustomNavi Container", for: .normal)
    button.addTarget(self, action: #selector(pushBenefitHubCustomNaviContainerViewController), for: .touchUpInside)
    return button
  }()

  private lazy var bannerButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Banner", for: .normal)
    button.addTarget(self, action: #selector(pushBannerViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var inquiryButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Inquiry", for: .normal)
    button.addTarget(self, action: #selector(showInquiry), for: .touchUpInside)
    return button
  }()
  
  private lazy var privacyConsentStatusLabel: UILabel = {
    let label = UILabel(frame: .zero)
    label.text = "Before Load Status"
    label.numberOfLines = 0
    
    return label
  }()
  
  private lazy var loadPrivacyConsentStatusButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Load PrivacyConsent Status", for: .normal)
    button.addTarget(self, action: #selector(loadPrivacyConsentStatus), for: .touchUpInside)
    
    return button
  }()
  
  private lazy var grantPrivacyConsentButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Grant PrivacyConsent", for: .normal)
    button.addTarget(self, action: #selector(grantPrviacyConsent), for: .touchUpInside)
    
    return button
  }()

  override func viewDidLoad() {
    super.viewDidLoad()
    
    setupView()
    setupLayout()
  }
  
  private func setupView() {
    view.backgroundColor = .systemBackground

    navigationItem.title = "BuzzvilSDK-Swift"
    
    rootStackView.addArrangedSubview(nativeButton)
    rootStackView.addArrangedSubview(interstitialButton)
    rootStackView.addArrangedSubview(carouselButton)
    rootStackView.addArrangedSubview(benefitHubButton)
    rootStackView.addArrangedSubview(benefitHubContainerButton)
    rootStackView.addArrangedSubview(benefitHubCustomNaviContainerButton)
    rootStackView.addArrangedSubview(bannerButton)
    rootStackView.addArrangedSubview(inquiryButton)
    rootStackView.addArrangedSubview(privacyConsentStatusLabel)
    rootStackView.addArrangedSubview(loadPrivacyConsentStatusButton)
    rootStackView.addArrangedSubview(grantPrivacyConsentButton)

    view.addSubview(rootStackView)
    
    let buttonsToSetupAppearance = [
      nativeButton,
      interstitialButton,
      carouselButton,
      benefitHubButton,
      benefitHubContainerButton,
      benefitHubCustomNaviContainerButton,
      bannerButton,
      inquiryButton,
      loadPrivacyConsentStatusButton,
      grantPrivacyConsentButton,
    ]
    
    buttonsToSetupAppearance.forEach { button in
      setupAppearanceOfButton(button: button)
    }
  }
  
  private func setupLayout() {
    rootStackView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      rootStackView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
      rootStackView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
      rootStackView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 8),
      rootStackView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -8),
    ])
  }
  
  private func setupAppearanceOfButton(button: UIButton) {
    button.backgroundColor = .systemBlue
    button.layer.cornerRadius = 8
    button.setTitleColor(.white, for: .normal)
    button.titleLabel?.font = UIFont.systemFont(ofSize: 16)
    button.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      button.heightAnchor.constraint(equalToConstant: 32)
    ])
  }
  
  @objc private func pushNativeAdViewController() {
    let nativeAdViewController = NativeViewController()
    navigationController?.pushViewController(nativeAdViewController, animated: true)
  }
  
  @objc private func pushCarouselViewController() {
    let carouselConfigurationViewController = CarouselViewController()
    navigationController?.pushViewController(carouselConfigurationViewController, animated: true)
  }
  
  @objc private func pushInterstitialViewController() {
    let interstitialViewController = InterstitialViewController()
    navigationController?.pushViewController(interstitialViewController, animated: true)
  }
  
  @objc private func pushBenefitHubViewController() {
    let benefitHubViewController = BenefitHubViewController()
    navigationController?.pushViewController(benefitHubViewController, animated: true)
  }

  @objc private func pushBenefitHubContainerViewController() {
    let benefitHubContainerViewController = BenefitHubContainerViewController()
    navigationController?.pushViewController(benefitHubContainerViewController, animated: true)
  }
  
  @objc private func pushBenefitHubCustomNaviContainerViewController() {
    let benefitHubCustomNaviViewController = BenefitHubNavigationCustomContainerViewController()
    navigationController?.pushViewController(benefitHubCustomNaviViewController, animated: true)
  }

  @objc private func pushBannerViewController() {
    let bannerViewController = BannerViewController()
    navigationController?.pushViewController(bannerViewController, animated: true)
  }
  
  @objc private func showInquiry() {
    BuzzAdBenefit.shared.openInquiryPage(unitId: "YOUR_UNIT_ID")
  }
  
  @objc private func loadPrivacyConsentStatus() {
    BuzzAdBenefit.shared.loadPrivacyConsentStatus { [weak self] status in
      guard let self = self else { return }
      switch status {
      case .granted:
        self.privacyConsentStatusLabel.text = "Granted"
      case .revoked:
        self.privacyConsentStatusLabel.text = "Revoked"
      }
    } onFailure: { [weak self] error in
      guard let self = self else { return }
      
      self.privacyConsentStatusLabel.text = "LoadPrivacyConsentStatus is failed Error: \(error)"
    }
  }
  
  @objc private func grantPrviacyConsent() {
    BuzzAdBenefit.shared.grantPrivacyConsent { [weak self] in
      guard let self = self else { return }
      self.privacyConsentStatusLabel.text = "Granted"
    } onFailure: { [weak self] error in
      guard let self = self else { return }
      
      self.privacyConsentStatusLabel.text = "GrantPrivacyConsentStatus is failed Error: \(error)"
    }
  }
}
